src/main/java/com/aws/greengrass/device/iot/IotAuthClient.java

```diff
@@ -55,29 +69,17 @@ public Optional<String> getActiveCertificateId(String certificatePem) {
                 throw new IllegalArgumentException("Certificate PEM is empty");
             }
 
-            VerifyClientDeviceIdentityRequest request =
-                    VerifyClientDeviceIdentityRequest.builder().clientDeviceCertificate(certificatePem).build();
-            try {
-                VerifyClientDeviceIdentityResponse response = RetryUtils.runWithRetry(SERVICE_EXCEPTION_RETRY_CONFIG,
-                        () -> clientFactory.getGreengrassV2DataClient().verifyClientDeviceIdentity(request),
-                        "verify-client-device-identity", logger);
-                return Optional.of(response.clientDeviceCertificateId());
-            } catch (InterruptedException e) {
-                logger.atError().cause(e).log("Verify client device identity got interrupted");
-                // interrupt the current thread so that higher-level interrupt handlers can take care of it
-                Thread.currentThread().interrupt();
-                throw new CloudServiceInteractionException(
-                        "Failed to verify client device identity, process got interrupted", e);
-            } catch (ValidationException | ResourceNotFoundException e) {
-                logger.atWarn().cause(e).kv("certificatePem", certificatePem)
-                        .log("Certificate doesn't exist or isn't active");
-                return Optional.empty();
-            } catch (Exception e) {
-                logger.atError().cause(e).kv("certificatePem", certificatePem)
-                        .log("Failed to verify client device identity with cloud. Check that the core device's IoT "
-                                + "policy grants the greengrass:VerifyClientDeviceIdentity permission.");
-                throw new CloudServiceInteractionException("Failed to verify client device identity", e);
+            // cache active IoT Certificate Ids locally to avoid multiple cloud requests;
+            // mapping of certificateHash <-> IoT Certificate Id is cached;
+            // certificateHash is SHA-256 hash of certificatePem (to avoid storing actual certificatePem)
+            String certId =  getCertificateIdAvailableLocally(certificatePem)
```

Comment:
nit: since we return optional , can just keep optional here instead of converting back to optional at the end

```
Optional<String> certId = ...
certId.ifPresent(id -> attachCertificateIdToPem(id, certificatePem))
return certId;

```

Reply:
Sorry couldn't find a way to chain `Optional` returns here. We can use `Optional<T>.or(() -> ...)` but that was added after JDK-9.

Reply:
makes sense now :) 
Updated.